src/maggma/stores/azure.py

```diff
@@ -0,0 +1,588 @@
+        """
+
+        if self.container is None or self.service is None:
+            raise RuntimeError("The store has not been connected")
+
+        if not isinstance(docs, list):
+            docs = [docs]
+
+        if isinstance(key, str):
+            key = [key]
+        elif not key:
+            key = [self.key]
+
+        if additional_metadata is None:
+            additional_metadata = []
+        elif isinstance(additional_metadata, str):
+            additional_metadata = [additional_metadata]
+        else:
+            additional_metadata = list(additional_metadata)
+
+        with ThreadPoolExecutor(max_workers=self.workers) as pool:
+            fs = {
+                pool.submit(
+                    self.write_doc_to_blob,
+                    doc=itr_doc,
+                    search_keys=key + additional_metadata + self.searchable_fields,
+                )
+                for itr_doc in docs
+            }
+            fs, _ = wait(fs)
+
+            search_docs = [sdoc.result() for sdoc in fs]
+
+        # Use store's update to remove key clashes
+        self.index.update(search_docs, key=self.key)
+
+    def _get_service_client(self):
+        if not hasattr(self._thread_local, "container"):
+            if isinstance(self.azure_client_info, str):
+                # assume it is the account_url and that the connection is passwordless
+                default_credential = DefaultAzureCredential()
+                return BlobServiceClient(
+                    self.azure_client_info, credential=default_credential
+                )
+
+            elif isinstance(self.azure_client_info, dict):
+                connection_string = self.azure_client_info.get("connection_string")
+                if connection_string:
+                    return BlobServiceClient.from_connection_string(
+                        conn_str=connection_string
+                    )
+
+            msg = f"Could not instantiate BlobServiceClient from azure_client_info: {self.azure_client_info}"
+            raise RuntimeError(msg)
+
+    def _get_container(self) -> Optional[ContainerClient]:
+        """
+        If on the main thread return the container created above, else create a new
+        container on each thread.
+        """
+        if threading.current_thread().name == "MainThread":
+            return self.container
+        if not hasattr(self._thread_local, "container"):
+            service_client = self._get_service_client()
+            container = service_client.get_container_client(self.container_name)
+            self._thread_local.container = container
+        return self._thread_local.container
+
+    def write_doc_to_blob(self, doc: Dict, search_keys: List[str]):
+        """
+        Write the data to an Azure blob and return the metadata to be inserted into the index db
+
+        Args:
+            doc: the document
+            search_keys: list of keys to pull from the docs and be inserted into the
```

Comment:
This is also the same from aws of course, from what I understand, search_keys can only be at first level right ? I would not change that now in any case, but wouldn't that be useful ?

Reply:
might be useful indeed. I agree not to change it now